File: 剑指Offer/32.从上到下打印二叉树2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def levelOrder(self, root: TreeNode) -> List[List[int]]:
        if not root:
            return []
        
        res = []
        singleLayer = []
        q = [root]
        toBeStored = 1
        nextLayer = 0
        while len(q) != 0:
            temp = q[0]
            if (temp.left != None):
                q.append(temp.left)
                nextLayer += 1
            if (temp.right != None):
                q.append(temp.right)
                nextLayer += 1
                
            q.pop(0)
            toBeStored -= 1
            singleLayer.append(temp.val)

            if toBeStored == 0:
                toBeStored = nextLayer
                nextLayer = 0
                res.append(singleLayer)
                singleLayer = []

        return res

File: 剑指Offer/32.从上到下打印二叉树2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        vector<vector<int> > res;
        vector<int> singleLayer;
        if (root == NULL) return res;
        queue<TreeNode*> q;
        q.push(root);
        int toBeStored = 1;
        int nextLayer = 0;
        while (!q.empty()) {
            TreeNode* temp = q.front();
            if (temp->left != NULL) {
                q.push(temp->left);
                nextLayer++;
            }
            if (temp->right != NULL) {
                q.push(temp->right);
                nextLayer++;
            }
            q.pop();
            singleLayer.push_back(temp->val);
            toBeStored--;
            if (toBeStored == 0) {
                toBeStored = nextLayer;
                nextLayer = 0;
                res.emplace_back(singleLayer);
                singleLayer.clear();
            }
        }

        return res;
    }
};